File: src/sbml/test/TestTrigger.c
/**
 * \file    TestTrigger.c
 * \brief   SBML Trigger unit tests
 * \author  Sarah Keating
 *
 * $Id$
 * $Source$
 */
/* Copyright 2003 California Institute of Technology and
 * Japan Science and Technology Corporation.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2.1 of the License, or
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  The software and
 * documentation provided hereunder is on an "as is" basis, and the
 * California Institute of Technology and Japan Science and Technology
 * Corporation have no obligations to provide maintenance, support,
 * updates, enhancements or modifications.  In no event shall the
 * California Institute of Technology or the Japan Science and Technology
 * Corporation be liable to any party for direct, indirect, special,
 * incidental or consequential damages, including lost profits, arising
 * out of the use of this software and its documentation, even if the
 * California Institute of Technology and/or Japan Science and Technology
 * Corporation have been advised of the possibility of such damage.  See
 * the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *
 * The original code contained here was initially developed by:
 *
 *     Ben Bornstein
 *     The Systems Biology Markup Language Development Group
 *     ERATO Kitano Symbiotic Systems Project
 *     Control and Dynamical Systems, MC 107-81
 *     California Institute of Technology
 *     Pasadena, CA, 91125, USA
 *
 *     http://www.cds.caltech.edu/erato
 *     mailto:sbml-team@caltech.edu
 *
 * Contributor(s):
 */


#include "common/common.h"
#include "math/FormulaParser.h"

#include "SBase.h"
#include "Trigger.h"

#include <check.h>


static Trigger_t *D;


void
TriggerTest_setup (void)
{
  D = Trigger_create();

  if (D == NULL)
  {
    fail("Trigger_create() returned a NULL pointer.");
  }
}


void
TriggerTest_teardown (void)
{
  Trigger_free(D);
}


START_TEST (test_Trigger_create)
{
  fail_unless( SBase_getTypeCode((SBase_t *) D) == SBML_TRIGGER );
  fail_unless( SBase_getMetaId    ((SBase_t *) D) == NULL );
  fail_unless( SBase_getNotes     ((SBase_t *) D) == NULL );
  fail_unless( SBase_getAnnotation((SBase_t *) D) == NULL );

  fail_unless( Trigger_getMath(D) == NULL );
}
END_TEST


START_TEST (test_Trigger_createWithMath)
{
  ASTNode_t            *math = SBML_parseFormula("x^3");
  Trigger_t *fd   = Trigger_createWithMath(math);

  ASTNode_t * math1;
  char * formula;

  fail_unless( SBase_getTypeCode((SBase_t *) fd) == SBML_TRIGGER );
  fail_unless( SBase_getMetaId    ((SBase_t *) fd) == NULL );
  fail_unless( SBase_getNotes     ((SBase_t *) fd) == NULL );
  fail_unless( SBase_getAnnotation((SBase_t *) fd) == NULL );


  math1 = Trigger_getMath(fd);
  fail_unless( math1 != NULL );

  formula = SBML_formulaToString(math1);
  fail_unless( formula != NULL );
  fail_unless( !strcmp(formula, "x^3") );
  fail_unless( Trigger_getMath(fd) != math );
  fail_unless( Trigger_isSetMath(fd) );


  Trigger_free(fd);
}
END_TEST


START_TEST (test_Trigger_free_NULL)
{
  Trigger_free(NULL);
}
END_TEST


START_TEST (test_Trigger_setMath)
{
  ASTNode_t *math = SBML_parseFormula("lambda(x, x^3)");

  ASTNode_t * math1;
  char * formula;

  Trigger_setMath(D, math);

  math1 = Trigger_getMath(D);
  fail_unless( math1 != NULL );

  formula = SBML_formulaToString(math1);
  fail_unless( formula != NULL );
  fail_unless( !strcmp(formula, "lambda(x, x^3)") );
  fail_unless( Trigger_getMath(D) != math );
  fail_unless( Trigger_isSetMath(D) );

  /* Reflexive case (pathological) */
  Trigger_setMath(D, (ASTNode_t *) Trigger_getMath(D));
  math1 = Trigger_getMath(D);
  fail_unless( math1 != NULL );

  formula = SBML_formulaToString(math1);
  fail_unless( formula != NULL );
  fail_unless( !strcmp(formula, "lambda(x, x^3)") );

  Trigger_setMath(D, NULL);
  fail_unless( !Trigger_isSetMath(D) );

  if (Trigger_getMath(D) != NULL)
  {
    fail("Trigger_setMath(D, NULL) did not clear ASTNode.");
  }
}
END_TEST


Suite *
create_suite_Trigger (void)
{
  Suite *suite = suite_create("Trigger");
  TCase *tcase = tcase_create("Trigger");


  tcase_add_checked_fixture( tcase,
                             TriggerTest_setup,
                             TriggerTest_teardown );

  tcase_add_test( tcase, test_Trigger_create       );
  tcase_add_test( tcase, test_Trigger_createWithMath   );
  tcase_add_test( tcase, test_Trigger_setMath      );
  tcase_add_test( tcase, test_Trigger_free_NULL );

  suite_add_tcase(suite, tcase);

  return suite;
}

File: src/bindings/python/test/TestTrigger.py
#
# @file    TestTrigger.py
# @brief   SBML Trigger unit tests
# @author  Akiya Jouraku (Python conversion)
# @author  Sarah Keating 
#
# $Id$
# $Source$
#
# This test file was converted from src/sbml/test/TestTrigger.c
# with the help of conversion sciprt (ctest_converter.pl).
#
#<!---------------------------------------------------------------------------
# This file is part of libSBML.  Please visit http://sbml.org for more
# information about SBML, and the latest version of libSBML.
#
# Copyright 2005-2008 California Institute of Technology.
# Copyright 2002-2005 California Institute of Technology and
#                     Japan Science and Technology Corporation.
# 
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation.  A copy of the license agreement is provided
# in the file named "LICENSE.txt" included with this software distribution
# and also available online as http://sbml.org/software/libsbml/license.html
#--------------------------------------------------------------------------->*/
#
import sys
import unittest
import libsbml

class TestTrigger(unittest.TestCase):

  D = None

  def setUp(self):
    self.D = libsbml.Trigger()
    if (self.D == None):
      pass    
    pass  

  def tearDown(self):
    self.D = None
    pass  

  def test_Trigger_create(self):
    self.assert_( self.D.getTypeCode() == libsbml.SBML_TRIGGER )
    self.assert_( self.D.getMetaId() == "" )
    self.assert_( self.D.getNotes() == None )
    self.assert_( self.D.getAnnotation() == None )
    self.assert_( self.D.getMath() == None )
    pass  

  def test_Trigger_createWithMath(self):
    math = libsbml.parseFormula("x^3")
    fd = libsbml.Trigger(math)
    self.assert_( fd.getTypeCode() == libsbml.SBML_TRIGGER )
    self.assert_( fd.getMetaId() == "" )
    self.assert_( fd.getNotes() == None )
    self.assert_( fd.getAnnotation() == None )
    math1 = fd.getMath()
    self.assert_( math1 != None )
    formula = libsbml.formulaToString(math1)
    self.assert_( formula != None )
    self.assert_((  "x^3" == formula ))
    self.assert_( fd.getMath() != math )
    self.assertEqual( True, fd.isSetMath() )
    fd = None
    pass  

  def test_Trigger_free_NULL(self):
    
    pass  

  def test_Trigger_setMath(self):
    math = libsbml.parseFormula("lambda(x, x^3)")
    self.D.setMath(math)
    math1 = self.D.getMath()
    self.assert_( math1 != None )
    formula = libsbml.formulaToString(math1)
    self.assert_( formula != None )
    self.assert_((  "lambda(x, x^3)" == formula ))
    self.assert_( self.D.getMath() != math )
    self.assertEqual( True, self.D.isSetMath() )
    self.D.setMath(self.D.getMath())
    math1 = self.D.getMath()
    self.assert_( math1 != None )
    formula = libsbml.formulaToString(math1)
    self.assert_( formula != None )
    self.assert_((  "lambda(x, x^3)" == formula ))
    self.D.setMath(None)
    self.assertEqual( False, self.D.isSetMath() )
    if (self.D.getMath() != None):
      pass    
    pass  

def suite():
  suite = unittest.TestSuite()
  suite.addTest(unittest.makeSuite(TestTrigger))

  return suite

if __name__ == "__main__":
  if unittest.TextTestRunner(verbosity=1).run(suite()).wasSuccessful() :
    sys.exit(0)
  else:
    sys.exit(1)
